(2022.10.05) Stefan Verkoyen (sverkoyen.contractor@libertyglobal.com)
ONEM-22741 [dev] LISA locking mechanism
ONEM-22742 [dev] LISA API Extension: Lock
ONEM-22743 [dev] LISA API Extension: Unlock
ONEM-22744 [dev] LISA API Extension: notifications/operationStatus

* possibility to lock/unlock an app
* locked apps cannot be uninstalled (and cannot be locked
  more than once at a time)
* operationStatus events contain more info:
  - type,id,version of app
  - operation (Installing,Uninstalling)
  - status: new "Cancelled" status next to Success, Failed and
    Progress.

diff --git a/interfaces/ILISA.h b/interfaces/ILISA.h
index 080e5b8..ceaec66 100644
--- a/interfaces/ILISA.h
+++ b/interfaces/ILISA.h
@@ -136,9 +136,8 @@ namespace Exchange {
             enum {ID = ID_LISA_NOTIFICATION};
 
             /* @brief Completion of asynchronous operation. */
-            virtual void operationStatus(const std::string& handle,
-                    const std::string& status,
-                    const std::string& details) = 0;
+            virtual void operationStatus(const string& handle, const string& operation, const string& type, const string& id,
+                                         const string& version, const string& status, const string& details) = 0;
         };
 
         virtual uint32_t Register(ILISA::INotification* notification) = 0;
@@ -194,6 +193,37 @@ namespace Exchange {
                 const std::string& appName,
                 const std::string& category,
                 IAppsPayload*& result /* @out */) const = 0;
+
+        struct EXTERNAL IHandleResult : virtual public Core::IUnknown {
+            enum { ID = ID_LISA_HANDLE_RESULT };
+
+            virtual uint32_t Handle(std::string& handle /* @out */) const = 0;
+        };
+
+        /* @brief Lock the application. Preventing uninstallation. */
+        virtual uint32_t Lock(const std::string& type,
+                const std::string& id,
+                const std::string& version,
+                const std::string& reason,
+                const std::string& owner,
+                ILISA::IHandleResult*& result /* @out */) = 0;
+
+
+        /* @brief Unlock application. */
+        virtual uint32_t Unlock(const std::string& handle) = 0;
+
+        struct EXTERNAL ILockInfo : virtual public Core::IUnknown {
+            enum { ID = ID_LISA_LOCK_INFO };
+
+            virtual uint32_t Reason(std::string& reason /* @out */) const = 0;
+            virtual uint32_t Owner(std::string& owner /* @out */) const = 0;
+        };
+
+        /* @brief Get lock info. */
+        virtual uint32_t GetLockInfo(const std::string& type,
+                const std::string& id,
+                const std::string& version,
+                ILISA::ILockInfo*& result /* @out */) = 0;
     };
 
 }
diff --git a/interfaces/Ids.h b/interfaces/Ids.h
index fe8193e..68805d8 100644
--- a/interfaces/Ids.h
+++ b/interfaces/Ids.h
@@ -229,6 +229,8 @@ namespace Exchange {
         ID_LISA_KEY_VALUE,
         ID_LISA_KEY_VALUE_ITERATOR,
         ID_LISA_METADATA_PAYLOAD,
+        ID_LISA_LOCK_INFO,
+        ID_LISA_HANDLE_RESULT,
 
         ID_CONTENTDECRYPTION_NOTIFICATION,
     };
diff --git a/jsonrpc/LISA.json b/jsonrpc/LISA.json
index 97670e4..f398f4a 100644
--- a/jsonrpc/LISA.json
+++ b/jsonrpc/LISA.json
@@ -9,6 +9,32 @@
   "common": {
     "$ref": "common.json"
   },
+  "errors": {
+    "wrongparams": {
+      "code": 1001,
+      "message": "ERROR_WRONG_PARAMS"
+    },
+    "toomanyrequests": {
+      "code": 1002,
+      "message": "ERROR_TOO_MANY_REQUESTS"
+    },
+    "alreadyinstalled": {
+      "code": 1003,
+      "message": "ERROR_ALREADY_INSTALLED"
+    },
+    "wronghandle": {
+      "code": 1007,
+      "message": "ERROR_WRONG_HANDLE"
+    },
+    "applocked": {
+      "code": 1009,
+      "message": "ERROR_APP_LOCKED"
+    },
+    "appuninstalling": {
+      "code": 1010,
+      "message": "ERROR_APP_UNINSTALLING"
+    }
+  },
   "definitions": {
     "appId": {
       "description": "Application ID",
@@ -44,9 +70,35 @@
       "description": "Operation handle",
       "type": "string"
     },
+    "handleResult": {
+      "type": "object",
+      "description": "Struct holding handle",
+      "properties": {
+        "handle": {
+          "$ref": "#/definitions/handle"
+        }
+      },
+      "required": [
+        "handle"
+      ]
+    },
+    "emptyResult": {
+      "type": "object",
+      "description": "Struct holding nothing i.e. {}",
+      "properties": {
+      },
+      "required": [
+      ]
+    },
     "status": {
       "description": "Operation status",
-      "type": "string"
+      "type": "string",
+      "example": "Success"
+    },
+    "operation": {
+      "description": "Operation",
+      "type": "string",
+      "example": "Installing"
     },
     "progress": {
       "description": "Operation progress",
@@ -60,6 +112,30 @@
       "description": "Value property",
       "type": "string"
     },
+    "reason": {
+      "description": "Reason for lock",
+      "type": "string",
+      "example": "active"
+    },
+    "owner": {
+      "description": "Owner of lock",
+      "type": "string",
+      "example": "awc"
+    },
+    "lockInfo": {
+      "type": "object",
+      "description": "Struct holding lock info",
+      "properties": {
+        "reason": {
+          "$ref": "#/definitions/reason"
+        },
+        "owner": {
+          "$ref": "#/definitions/owner"
+        }
+      },
+      "required": [
+      ]
+    },
     "installedApp": {
       "type": "object",
       "description": "Installed application variant",
@@ -681,6 +757,114 @@
           "$ref": "#/common/errors/general"
         }
       ]
+    },
+    "lock": {
+      "summary": "Lock an application",
+      "params": {
+        "type": "object",
+        "properties": {
+          "type": {
+            "$ref": "#/definitions/appType"
+          },
+          "id": {
+            "$ref": "#/definitions/appId"
+          },
+          "version": {
+            "$ref": "#/definitions/appVersion"
+          },
+          "reason": {
+            "$ref": "#/definitions/reason"
+          },
+          "owner": {
+            "$ref": "#/definitions/owner"
+          }
+        },
+        "required": [
+          "id",
+          "type",
+          "version"
+        ]
+      },
+      "result": {
+        "$ref": "#/definitions/handleResult"
+      },
+      "errors": [
+        {
+          "description": "LISA initialization in progress or other operation pending",
+          "$ref": "#/errors/toomanyrequests"
+        },
+        {
+          "description": "Invalid params or app not unstalled",
+          "$ref": "#/errors/wrongparams"
+        },
+        {
+          "description": "App is currently being uninstalled",
+          "$ref": "#/errors/appuninstalling"
+        },
+        {
+          "description": "App is already locked",
+          "$ref": "#/errors/applocked"
+        }
+      ]
+    },
+    "unlock": {
+      "summary": "Unlock an application",
+      "params": {
+        "type": "object",
+        "$ref": "#/common/results/void",
+        "properties": {
+          "handle": {
+            "$ref": "#/definitions/handle"
+          }
+        },
+        "required": [
+          "handle"
+        ]
+      },
+      "result": {
+        "$ref": "#/definitions/emptyResult"
+      },
+      "errors": [
+        {
+          "description": "Wrong handle or app not locked anymore",
+          "$ref": "#/errors/wronghandle"
+        }
+      ]
+    },
+    "getLockInfo": {
+      "summary": "Get lock info",
+      "params": {
+        "type": "object",
+        "properties": {
+          "type": {
+            "$ref": "#/definitions/appType"
+          },
+          "id": {
+            "$ref": "#/definitions/appId"
+          },
+          "version": {
+            "$ref": "#/definitions/appVersion"
+          }
+        },
+        "required": [
+          "id",
+          "type",
+          "version"
+        ]
+      },
+      "result": {
+        "$ref": "#/definitions/lockInfo"
+      },
+      "errors": [
+        {
+          "description": "Invalid params or app not unstalled",
+          "$ref": "#/errors/wrongparams"
+        },
+        {
+          "description": "Wrong handle or app not locked anymore",
+          "$ref": "#/errors/wronghandle"
+        }
+      ]
     }
   },
   "events": {
@@ -693,6 +877,18 @@
           "handle": {
             "$ref": "#/definitions/handle"
           },
+          "operation": {
+            "$ref": "#/definitions/operation"
+          },
+          "type": {
+            "$ref": "#/definitions/appType"
+          },
+          "id": {
+            "$ref": "#/definitions/appId"
+          },
+          "version": {
+            "$ref": "#/definitions/appVersion"
+          },
           "status": {
             "$ref": "#/definitions/status"
           },
@@ -703,7 +899,11 @@
         },
         "required": [
           "handle",
-          "status"
+          "status",
+          "operation",
+          "id",
+          "type",
+          "version"
         ]
       }
     }
